apollo-federation/src/supergraph/mod.rs

```diff
@@ -1693,103 +1653,6 @@ lazy_static! {
     };
 }
 
-fn insert_directive(
-    schema: &mut FederationSchema,
-    pos: &CompositeTypeDefinitionPosition,
-    directive: Component<Directive>,
-) -> Result<(), FederationError> {
-    match pos {
-        CompositeTypeDefinitionPosition::Union(pos) => pos.insert_directive(schema, directive),
-        CompositeTypeDefinitionPosition::Object(pos) => pos.insert_directive(schema, directive),
-        CompositeTypeDefinitionPosition::Interface(pos) => pos.insert_directive(schema, directive),
-    }
-}
-
-trait CompositeType {
-    fn directives(&self) -> &DirectiveList;
-}
-
-impl CompositeType for UnionType {
-    fn directives(&self) -> &DirectiveList {
-        &self.directives
-    }
-}
-
-impl CompositeType for ObjectType {
-    fn directives(&self) -> &DirectiveList {
-        &self.directives
-    }
-}
-
-impl CompositeType for InterfaceType {
-    fn directives(&self) -> &DirectiveList {
-        &self.directives
-    }
-}
-
-fn apply_context_to_type<T>(
```

Comment:
Refactors:
- This was moved into empty type addition, similar to the JS code. This allows the code to be called once.